### PR TITLE
[rfc][doc] Add clarity to stability guidelines

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Google guidelines were 1. not indexable and 2. not precisely relevant for our project.

This PR aims to make the stability guidelines much more clear.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-contribute/getting-involved.rst [hunk 1]
```diff
@@ -246,13 +246,67 @@ check recent tests known to be flaky.
 
 .. _`CI`: https://github.com/ray-project/ray/tree/master/ci
 
+.. _api-stability:
```
Comment (richardliaw) [inline doc/source/ray-contribute/getting-involved.rst:249]: should we make this into a separate page?

Comment (ericl) [reply doc/source/ray-contribute/getting-involved.rst:249]: +1

File: doc/source/ray-contribute/getting-involved.rst [hunk 2]
```diff
@@ -246,13 +246,67 @@ check recent tests known to be flaky.
...
-Ray provides stability guarantees for its public APIs in Ray core and libraries. The level of stability provided depends on how the API is annotated.
+Ray provides stability guarantees for its public APIs in Ray core and libraries, which are decorated/labeled accordingly.
+
+An API can be labeled:
+
+* :ref:`PublicAPI <public-api-def>`, which means the API is exposed to end users. PublicAPI has three sub-levels (alpha, beta, stable), as described below.
+* :ref:`DeveloperAPI <developer-api-def>`, which means the API is explicitly exposed to *advanced* Ray users and library developers
+* :ref:`Deprecated <deprecated-api-def>`, which may be removed in future releases of Ray.
+
+Ray's PublicAPI stability definitions are based off the `Google stability level guidelines <https://google.aip.dev/181>`_, with minor differences:
```
Comment (ericl) [inline doc/source/ray-contribute/getting-involved.rst:262]: Should we also unify with RLlib stability annotations? (delete that page / direct here)

File: python/ray/util/annotations.py [hunk 3]
```diff
@@ -44,10 +45,15 @@ def wrap(obj):
...
-                "and may change before becoming stable."
+                "and may change before becoming stable. "
             )
         else:
             obj.__doc__ += "\n    PublicAPI: This API is stable across Ray releases."
+
+        obj.__doc__ += (
+            "(`Definitions <https://docs.ray.io/en/latest/"
+            "ray-contribute/getting-involved.html#api-stability>`_)"
+        )
```
Comment (ericl) [inline python/ray/util/annotations.py:56]: This seems a bit much, can we omit it?

**Review by ericl — CHANGES_REQUESTED** [review]

File: doc/source/ray-contribute/getting-involved.rst (hunk 2 above)
Comment (richardliaw) [reply doc/source/ray-contribute/getting-involved.rst]: I've created an issue for Sven/team to migrate their annotations #25628 .

File: doc/source/ray-contribute/getting-involved.rst (hunk 1 above)
Comment (richardliaw) [reply doc/source/ray-contribute/getting-involved.rst]: done

File: python/ray/util/annotations.py [hunk 4]
```diff
@@ -44,10 +45,11 @@ def wrap(obj):
         if stability in ["alpha", "beta"]:
             obj.__doc__ += (
                 f"\n    PublicAPI ({stability}): This API is in {stability} "
-                "and may change before becoming stable."
+                "and may change before becoming stable. "
```
Comment (ericl) [inline python/ray/util/annotations.py:48]: Extra space.

**Review by ericl — COMMENTED** [review]
Please fix the extra space before merging.

**Review by ericl — APPROVED** [review]

File: python/ray/util/annotations.py (hunk 4 above)
Comment (richardliaw) [inline python/ray/util/annotations.py:48]:
```suggestion
                "and may change before becoming stable."
```

**Pull request merged**